### ISSUE TITLE
runProgram/findProgram wishlist

### ISSUE THREAD

**Issue by mahrud** [Macaulay2/M2#2762]
This is somehow the first time I've ever had to deal with these two methods, so here are some thoughts about what I expected before reading the documentation:
1. `normaliz = findProgram "normaliz"` should work (e.g. if I'm just prototyping)
2. `runProgram("normaliz", "--version")` should also work (e.g. if I'm just lazy)
3. `status runProgram(normaliz, "--version")` should give the return code
4. `toString runProgram(normaliz, "--version")` should give the string output
5. `normaliz << input` should do the obvious thing, which is to run with that input
6. `MinimumVersion` would be useless if I didn't know bash-fu. Why not take the minimum version and then a regex rule for getting the version from the output:
```m2
normaliz = findProgram("normaliz", "--version", "2.8", "Normaliz ([0-9.]+)")
```
Otherwise it would be simpler to just use `get "!normaliz --version"` (which is _impossible_ to find in the documentation unless you already know about it!). Alternatively, if the version has to be extracted from a file, I would expect something like:
```m2
ver = select("NMZ_VERSION *([0-9.]+)", "$1", get "libnormaliz/version.h")
normaliz = findProgram("normaliz", "--version", ver)
```
That is, if there are only three arguments, the third argument is the stated version, but if there are four arguments, the 3rd one is the minimum version and the 4th is the regex for extracting it.

cc: @d-torrance

**Comment by d-torrance** [comment]
Great suggestions!

**Comment by d-torrance** [comment]
Just to clarify -- is this the behavior you were suggesting for `Program << Thing`?

```m2
i1 : M2 = findProgram "M2"

o1 = M2

o1 : Program

i2 : M2 << "2 + 2"

o2 = 0

o2 : ProgramRun

i3 : toString oo

o3 = 
     i1 : 2 + 2
          
     o1 = 4
```

**Comment by mahrud** [comment]
Exactly! It may be too much syntactic sugar, but easy to remember syntax gets used more :) We can think about this longer also, because perhaps something like this would also be nice:
```m2
runProgram(M2, "-q") << "2 + 2"
```
but that might require non-superficial changes.

**Comment by mahrud** [comment]
Updates:
1. `msolve = findProgram "msolve"` doesn't work because running `msolve` alone prints help and returns code 1, which is not an uncommon choice. As long as the program is found and it is executable, I think this should return a `Program` object (i.e. the user chose not to provide a test command, so no need to test!)
2. after running the above, `msolve(args)` should just work.
3. moreover `msolve_args << stdinput` should also work.
4. running `findProgram("msolve", MinimumVersion => ("0.6.7", "msolve -V"), RaiseError => false)` returns null if the version is too low. I think it should still return a `Program` object, but either with a warning or `runProgram` should then give an error.
5. `program#"version"` should be `program.Version` perhaps.
6. `runProgram` should also accept a list/sequence/hashtable/optiontable of args and insert spaces.

**Comment by mahrud** [comment]
7. `runProgram` should have an option to display the output or at least the error live, rather than keeping it for the end. It's important to know that status of some program (e.g. msolve) if it's going to take hours/days.

@d-torrance do you think you could implement this one soon? I actually think without this I would prefer to not use runProgram in Msolve.

**Comment by d-torrance** [comment]
> **Written by Claude** (Claude Opus 5, via Claude Code), not by @d-torrance, whose account posted it -- please weigh it accordingly.

Six of the thirteen requests on this thread are already implemented, so here is the state of each on
1.26.06 — measured, not read off the documentation.

From the original list:

- [x] **1.** `findProgram "normaliz"` with no test command — works.
- [x] **2.** `runProgram("normaliz", "--version")` with a string name — works.
- [x] **3.** `status runProgram(p, "--version")` — returns the code.
- [x] **4.** `toString runProgram(p, "--version")` — returns the output.
- [x] **5.** `p << input` — works, and does what @d-torrance proposed in this thread.
- [ ] **6.** `MinimumVersion` with a regex for extracting the version. Neither the
  `MinimumVersion => (min, cmd)` tuple nor the four-argument
  `findProgram(name, cmd, min, regex)` form parses.

From the 2024 additions:

- [ ] **1.** `findProgram` still requires the bare program to exit 0, so the `msolve` case stands.
  `findProgram "false"` raises while `findProgram "true"` succeeds; `cat` and `M2` pass only because
  they happen to exit 0.
- [ ] **2.** `p args` directly — raises.
- [ ] **3.** `p_args << stdinput` — raises.
- [ ] **4.** `RaiseError => false` with too high a minimum still returns `null` rather than a
  `Program`.
- [ ] **5.** `p.Version` raises, and there is no version key at all: a `Program` found without
  `MinimumVersion` has keys `{path, prefix, name}`.
- [ ] **6.** `runProgram` with a list or sequence of arguments — raises.
- [x] **7.** Live output. Done in bf9ac3eade, 2024-08-01, *"Print verbose runProgram output as it
  happens"* — the same day it was asked for here, which is presumably why nobody said so. Use
  `Verbose => true`; the commit message notes stdout may still not appear until the end if the
  program does not flush.

For reference, `options runProgram` is `{RaiseError, RunDirectory, KeepFiles, Verbose}`.

<sub>Triaged in a sweep of the open issues; source of truth: [`bug-triage/issues.tsv`](https://github.com/d-torrance/M2/blob/bug-triage/bug-triage/issues.tsv).</sub>
